### PR TITLE
refactor(sdk)!: improve mock context provider async processing

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

Mock context provider seems to not work correctly (deadlock detected).

## What was done?

* Sdk::set_context_provider() no longer requires `&mut self`
* Created abstraction layer to execute async code in sync context (block_on() method)
* Refactored load_expectations as sync code
* Moved core_client to core module and exposed it publicly

## How Has This Been Tested?

GHA

Integrated with third-party software

## Breaking Changes

1. Sdk::set_context_provider no longer requires &mut self.
2. In Mock Sdk, load_expectations() is deprecated in favort of load_expectations_sync()

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [ ] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
	- Introduced a new error variant, `AsyncError`, for improved asynchronous error handling.
	- Added a new example, `read_contract`, to demonstrate usage with mocks.
	- Enhanced modularity with the addition of the `dash_core_client` module, conditionally compiled for testing.
	- Implemented utilities for handling asynchronous code within synchronous contexts.
	- Added `load_expectation` function for streamlined loading of expectations.

- **Bug Fixes**
	- Updated test cases for the `ContestedResource` object to improve robustness and error handling.

- **Refactor**
	- Renamed `CoreClient` to `LowLevelDashCoreClient` and updated related methods.
	- Split the `load_expectations` method into synchronous and asynchronous versions for better clarity.
	- Updated the `core_password` field type in the configuration for enhanced security.
	- Changed the `context_provider` type in `Sdk` for improved concurrency.

- **Chores**
	- Updated package versions and dependencies for better performance and compatibility.

### CODE REVIEW

**Review by QuantumExplorer — APPROVED** [review]

**Comment by lklimek** [comment]
@coderabbitai review

**Comment by lklimek** [comment]
@coderabbitai review

File: packages/rs-sdk/src/mock/provider.rs [hunk 1]
```diff
@@ -197,9 +198,9 @@ impl ContextProvider for GrpcContextProvider {
 
         let sdk_cloned = sdk.clone();
 
-        let data_contract: Option<DataContract> = DataContract::fetch(&sdk_cloned, contract_id)
-            .block_on()
-            .map_err(|e| ContextProviderError::DataContractFailure(e.to_string()))?;
+        let data_contract: Option<DataContract> =
+            block_on(async move { DataContract::fetch(&sdk_cloned, contract_id).await })?
+                .map_err(|e| ContextProviderError::DataContractFailure(e.to_string()))?;
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/mock/provider.rs:203]: Refactored, need to be re-checked.

File: packages/rs-sdk/src/sync.rs [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    tracing::trace!("block_on: Spawning worker");
+    let hdl = rt.spawn(worker(fut, tx));
+    tracing::trace!("block_on: Worker spawned");
+    let resp = tokio::task::block_in_place(|| rx.recv())?;
+
+    tracing::trace!("Response received");
+    if !hdl.is_finished() {
+        tracing::debug!("async-sync worker future is not finished, aborting; this should not happen, but it's fine");
+        hdl.abort(); // cleanup the worker future
+    }
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/sync.rs:71]: We cannot await for hdl, as we are not in async context. It means we cannot receive result of that thread in a clean way.

File: packages/rs-sdk/src/sync.rs [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    F: Future + Send + 'static,
+    F::Output: Send,
+{
+    tracing::trace!("block_on: running async function from sync code");
+    let rt = tokio::runtime::Handle::try_current()?;
+    let (tx, rx) = std::sync::mpsc::channel();
+    tracing::trace!("block_on: Spawning worker");
+    let hdl = rt.spawn(worker(fut, tx));
+    tracing::trace!("block_on: Worker spawned");
+    let resp = tokio::task::block_in_place(|| rx.recv())?;
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/sync.rs:65]: calling rx.recv() directly starves other threads.

File: packages/rs-sdk/src/mock/sdk.rs [hunk 4]
```diff
@@ -82,14 +83,25 @@ impl MockDashPlatformSdk {
...
+    /// Load all expectations from files in a directory asynchronously.
+    ///
+    /// See [MockDashPlatformSdk::load_expectations_sync()] for more details.
+    #[deprecated(since = "1.4.0", note = "use load_expectations_sync")]
+    pub async fn load_expectations<P: AsRef<std::path::Path> + Send + 'static>(
+        &mut self,
+        dir: P,
+    ) -> Result<&mut Self, Error> {
+        self.load_expectations_sync(dir)
+    }
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/mock/sdk.rs:95]: We leave it for backwards compatibility

File: packages/rs-sdk/src/mock/sdk.rs [hunk 5]
```diff
@@ -114,97 +126,80 @@
             .map(|f| f.path())
             .collect();
 
+        let mut dapi = block_on(self.dapi.clone().lock_owned())?;
+
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/mock/sdk.rs:130]: std::sync::MutexGuard is not Send, which is required inside the code base, as we use it also in async context.

**Review by shumkov — APPROVED** [review]

**Pull request merged**